### PR TITLE
.*: fix lint issues of not having comments for exported funcs and vars along with any remaining issues and enable remaining disabled rules

### PR DESCRIPTION
**Pull request by purnesh42H**
Addresses: #7444

RELEASE NOTES: None

### CODE REVIEW

File: balancer/endpointsharding/endpointsharding.go [hunk 1]
```diff
@@ -285,6 +285,8 @@ func (bw *balancerWrapper) UpdateState(state balancer.State) {
 	bw.es.updateState()
 }
 
+// ParseConfig parses a child config list and returns a LB config for the
+// gracefulswitch Balancer.
```
Comment (dfawley) [inline balancer/endpointsharding/endpointsharding.go:289]: Don't mention graceful switch here.  It's an implementation detail of this package.  It returns an LB config to use with the endpointsharding balancer.

But we can steal the description of how the config should look from gsb:

```
cfg is expected to be a JSON array of LB policy names + configs as the format of the loadBalancingConfig field in ServiceConfig.
```

File: credentials/tls/certprovider/pemfile/builder.go [hunk 2]
```diff
@@ -39,6 +40,8 @@ func init() {
 
 type pluginBuilder struct{}
 
+// ParseConfig parses the configuration data from JSON and returns a
+// certprovider.BuildableConfig for the PEM file watcher.
```
Comment (dfawley) [inline credentials/tls/certprovider/pemfile/builder.go:44]: Is vet failing for these?

It shouldn't.

This is an unexported type, so the docstrings won't appear anywhere.

These don't seem to provide any real value, so I'd just as soon delete them.  And I really hope we can configure the vet check better to not complain (if it is).

File: internal/channelz/server.go [hunk 3]
```diff
@@ -59,6 +59,8 @@ func NewServerMetricsForTesting(started, succeeded, failed, timestamp int64) *Se
 	return sm
 }
 
+// CopyFrom copies the metrics data from provided ServerMetrics
```
Comment (dfawley) [inline internal/channelz/server.go:62]: from _the_ provided...

File: internal/channelz/socket.go [hunk 4]
```diff
@@ -70,13 +70,23 @@ type EphemeralSocketMetrics struct {
 	RemoteFlowControlWindow int64
 }
 
+// SocketType represents the type of socket, indicating whether it is a
+// NormalSocket or a ListenSocket.
```
Comment (dfawley) [inline internal/channelz/socket.go:74]: No need to enumerate the types of sockets in this comment.

File: internal/channelz/socket.go [hunk 5]
```diff
@@ -70,13 +70,23 @@ type EphemeralSocketMetrics struct {
...
 type SocketType string
 
 const (
+	// SocketTypeNormal represents a standard socket used for normal
+	// operations.
```
Comment (dfawley) [inline internal/channelz/socket.go:79]: One comment above the const block should be sufficient.

https://tip.golang.org/doc/comment#const

File: internal/channelz/trace.go [hunk 6]
```diff
@@ -79,13 +79,21 @@ type TraceEvent struct {
...
-	EventNum     int64
-	mu           sync.Mutex
-	Events       []*traceEvent
+	// A counter for the number of events recorded in the
+	// trace.
+	EventNum int64
+	mu       sync.Mutex
+	// A slice of traceEvent pointers representing the events recorded for
+	// this channel.
+	Events []*traceEvent
```
Comment (dfawley) [inline internal/channelz/trace.go:96]: Hmm, should this be unexported? Its type is unexported, which is an anti-pattern.  It's probably unused outside this package would be my guess/hope.

File: internal/testutils/stats/test_metrics_recorder.go [hunk 7]
```diff
@@ -231,28 +258,37 @@ func (r *TestMetricsRecorder) RecordInt64Gauge(handle *estats.Int64GaugeHandle,
...
+// RecordFloat64Count is noop implementation of RecordFloat64Count.
 func (r *NoopMetricsRecorder) RecordFloat64Count(*estats.Float64CountHandle, float64, ...string) {}
 
+// RecordInt64Histo is noop implementation of RecordInt64Histo.
 func (r *NoopMetricsRecorder) RecordInt64Histo(*estats.Int64HistoHandle, int64, ...string) {}
 
+// RecordFloat64Histo is noop implementation of RecordFloat64Histo.
 func (r *NoopMetricsRecorder) RecordFloat64Histo(*estats.Float64HistoHandle, float64, ...string) {}
 
+// RecordInt64Gauge is moop implementation of RecordInt64Gauge.
```
Comment (dfawley) [inline internal/testutils/stats/test_metrics_recorder.go:293]: moop? :)

File: mem/buffers.go [hunk 8]
```diff
@@ -119,20 +122,25 @@ func Copy(data []byte, pool BufferPool) Buffer {
 	return NewBuffer(buf, pool)
 }
 
+// ReadOnlyData returns the underlying byte slice of the Buffer.
```
Comment (dfawley) [inline mem/buffers.go:125]: Similar as above.  This type is unexported; these methods should not need comments.

**Review by alialobidm — COMMENTED** [review]
- [x] [](url)

File: internal/channelz/trace.go (hunk 6 above)
Comment (purnesh42H) [reply internal/channelz/trace.go:96]: its being used in `channelz/internal/protoconv`

File: balancer/endpointsharding/endpointsharding.go (hunk 1 above)
Comment (purnesh42H) [reply balancer/endpointsharding/endpointsharding.go]: Removed and copied cfg part from downstream method

File: credentials/tls/certprovider/pemfile/builder.go (hunk 2 above)
Comment (purnesh42H) [reply credentials/tls/certprovider/pemfile/builder.go]: Removed docstring from all methods with unexported types

File: internal/channelz/server.go (hunk 3 above)
Comment (purnesh42H) [reply internal/channelz/server.go]: Done

File: internal/channelz/socket.go (hunk 4 above)
Comment (purnesh42H) [reply internal/channelz/socket.go]: Removed

File: internal/channelz/socket.go (hunk 5 above)
Comment (purnesh42H) [reply internal/channelz/socket.go]: Changed to `SocketType can be one of these.`

File: internal/transport/transport.go [hunk 9]
```diff
@@ -133,6 +133,14 @@ type recvBufferReader struct {
 	err         error
 }
 
+// ReadHeader reads data into the provided header slice. It first checks if
```
[the remark this answers is not in the document]
Comment (purnesh42H) [reply internal/transport/transport.go]: Removed as it is from unexported type

File: balancer/endpointsharding/endpointsharding.go (hunk 1 above)
Comment (dfawley) [reply balancer/endpointsharding/endpointsharding.go]: Can you use the text I suggested instead, please?

**Review by dfawley — CHANGES_REQUESTED** [review]

File: mem/buffers.go [hunk 10]
```diff
@@ -232,12 +236,21 @@ func (e emptyBuffer) read(buf []byte) (int, Buffer) {
...
+// methods for reading, splitting, and managing the byte slice.
 type SliceBuffer []byte
 
+// ReadOnlyData returns the byte slice.
 func (s SliceBuffer) ReadOnlyData() []byte { return s }
-func (s SliceBuffer) Ref()                 {}
-func (s SliceBuffer) Free()                {}
-func (s SliceBuffer) Len() int             { return len(s) }
+
+// Ref is noop implementation of Ref.
```
[the remark this answers is not in the document]
Comment (purnesh42H) [reply mem/buffers.go:246]: Done

File: mem/buffers.go [hunk 11]
```diff
@@ -211,12 +214,13 @@ func SplitUnsafe(buf Buffer, n int) (left, right Buffer) {
 	return buf.split(n)
 }
 
+// emptyBuffer is a Buffer implementation that represents an empty buffer. All
+// methods are no-op implementations.
```
[the remark this answers is not in the document]
Comment (purnesh42H) [reply mem/buffers.go:218]: Done

File: mem/buffers.go [hunk 12]
```diff
@@ -211,12 +214,13 @@ func SplitUnsafe(buf Buffer, n int) (left, right Buffer) {
...
 type emptyBuffer struct{}
 
 func (e emptyBuffer) ReadOnlyData() []byte {
 	return nil
 }
-
```
[the remark this answers is not in the document]
Comment (purnesh42H) [reply mem/buffers.go:219]: Done

File: scripts/vet.sh [hunk 13]
```diff
@@ -181,6 +181,6 @@ REV_OUT="$(mktemp)"
 revive -set_exit_status=1 -exclude "reflection/test/grpc_testing_not_regenerate/" -formatter plain ./... >"${REV_OUT}" || true
 
 # TODO: Remove `|| true` to unskip linter failures once existing issues are fixed.
-(not grep -v "\.pb\.go:" "${REV_OUT}") || true
+(noret_grep -v "reflection/test/grpc_testing_not_regenerate/" "${REV_OUT}" | not grep -v "\.pb\.go:") || true
```
Comment (arvindbr8) [inline scripts/vet.sh:184]: i already added this [here](https://github.com/purnesh42H/grpc-go/blob/master/scripts/vet.sh#L181)

Comment (purnesh42H) [reply scripts/vet.sh]: Oh ok. Thanks. Reverted

File: balancer/endpointsharding/endpointsharding.go [hunk 14]
```diff
@@ -285,6 +285,11 @@ func (bw *balancerWrapper) UpdateState(state balancer.State) {
 	bw.es.updateState()
 }
 
+// ParseConfig parses a child config list and returns a LB to use with the
```
Comment (dfawley) [inline balancer/endpointsharding/endpointsharding.go:288]: `returns a LoadBalancingConfig` please (or `returns an LB config` would be OK too).

File: internal/channelz/trace.go (hunk 6 above)
Comment (dfawley) [reply internal/channelz/trace.go:96]: That's not good.  We should export `traceEvent`, then.  If that's too hard to do in this PR, it's fine to leave it for now.

File: balancer/endpointsharding/endpointsharding.go (hunk 14 above)
Comment (purnesh42H) [reply balancer/endpointsharding/endpointsharding.go]: Done

File: internal/channelz/trace.go (hunk 6 above)
Comment (dfawley) [reply internal/channelz/trace.go:96]: That's not a very descriptive choice, but yes, some other name.

Is there any linter that covers unexported types being exposed outside a package, like this case?  If so I would like to turn it on.

**Review by dfawley — APPROVED** [review]

File: examples/features/advancedtls/client/main.go [hunk 15]
```diff
@@ -16,6 +16,9 @@
  *
  */
 
+// Binary client is an example client demonstrating use of advancedtls,
+// to set up a secure gRPC client connection with various TLS authentication
+// methods.
```
Comment (arvindbr8) [inline examples/features/advancedtls/client/main.go:21]: nit:

```suggestion
// Binary client is an example client demonstrating use of advancedtls, to set
// up a secure gRPC client connection with various TLS authentication methods.
```

File: internal/channelz/channel.go [hunk 16]
```diff
@@ -61,24 +65,29 @@ func (c *Channel) id() int64 {
...
 }
 
+// SubChans returns a copy of the map of sub-channels associated with the Channel.
 func (c *Channel) SubChans() map[int64]string {
 	db.mu.RLock()
 	defer db.mu.RUnlock()
 	return copyMap(c.subChans)
 }
 
+// NestedChans returns a copy of the map of nested channels associated with the Channel.
```
Comment (arvindbr8) [inline internal/channelz/channel.go:75]: pls wrap at 80 col

```suggestion
// NestedChans returns a copy of the map of nested channels associated with the
// Channel.
```

File: internal/channelz/channel.go [hunk 17]
```diff
@@ -61,24 +65,29 @@ func (c *Channel) id() int64 {
 	return c.ID
 }
 
+// SubChans returns a copy of the map of sub-channels associated with the Channel.
```
Comment (arvindbr8) [inline internal/channelz/channel.go:68]: pls wrap at 80 col:

```suggestion
// SubChans returns a copy of the map of sub-channels associated with the
// Channel.
```

File: internal/channelz/channel.go [hunk 18]
```diff
@@ -136,12 +145,16 @@ func strFromPointer(s *string) string {
 	return *s
 }
 
+// String returns a string representation of the ChannelMetrics, including its state,
+// target, and call metrics.
```
Comment (arvindbr8) [inline internal/channelz/channel.go:149]:
```suggestion
// String returns a string representation of the ChannelMetrics, including its
// state, target, and call metrics.
```

File: internal/channelz/channel.go [hunk 19]
```diff
@@ -136,12 +145,16 @@ func strFromPointer(s *string) string {
...
 func (c *ChannelMetrics) String() string {
 	return fmt.Sprintf("State: %v, Target: %s, CallsStarted: %v, CallsSucceeded: %v, CallsFailed: %v, LastCallStartedTimestamp: %v",
 		c.State.Load(), strFromPointer(c.Target.Load()), c.CallsStarted.Load(), c.CallsSucceeded.Load(), c.CallsFailed.Load(), c.LastCallStartedTimestamp.Load(),
 	)
 }
 
+// NewChannelMetricForTesting creates a new instance of ChannelMetrics with specified
+// initial values for testing purposes.
```
Comment (arvindbr8) [inline internal/channelz/channel.go:157]:
```suggestion
// NewChannelMetricForTesting creates a new instance of ChannelMetrics with
// specified initial values for testing purposes.
```

File: internal/channelz/server.go [hunk 20]
```diff
@@ -59,6 +59,8 @@ func NewServerMetricsForTesting(started, succeeded, failed, timestamp int64) *Se
 	return sm
 }
 
+// CopyFrom copies the metrics data from the provided ServerMetrics
+// instance into the current instance.
```
Comment (arvindbr8) [inline internal/channelz/server.go:63]:
```suggestion
// CopyFrom copies the metrics data from the provided ServerMetrics instance
// into the current instance.
```

File: internal/channelz/trace.go [hunk 21]
```diff
@@ -79,13 +79,21 @@ type TraceEvent struct {
 	Parent   *TraceEvent
 }
 
+// ChannelTrace provides tracing information for a channel.
+// It tracks various events and metadata related to the channel's lifecycle
+// and operations.
```
Comment (arvindbr8) [inline internal/channelz/trace.go:84]:
```suggestion
// ChannelTrace provides tracing information for a channel. It tracks various
// events and metadata related to the channel's lifecycle and operations.
```

File: internal/channelz/trace.go [hunk 22]
```diff
@@ -175,6 +183,7 @@ var refChannelTypeToString = map[RefChannelType]string{
 	RefNormalSocket: "NormalSocket",
 }
 
+// String returns a string representation of the RefChannelType
```
Comment (arvindbr8) [inline internal/channelz/trace.go:186]: Does the linter complain for Stringer method not having a comment? This seems unnecessary to me.

Comment (arvindbr8) [inline internal/channelz/trace.go:186]:
```suggestion
// String returns a string representation of the RefChannelType.
```

File: internal/idle/idle.go [hunk 23]
```diff
@@ -182,6 +182,7 @@ func (m *Manager) tryEnterIdleMode() bool {
 	return true
 }
 
+// EnterIdleModeForTesting instructs the channel to enter idle mode.
```
Comment (arvindbr8) [inline internal/idle/idle.go:185]: Very optional: 

Looking at https://pkg.go.dev/google.golang.org/grpc/internal/idle#Manager.EnterIdleModeForTesting, should we add more info from `tryEnterIdleMode()` here?

File: internal/idle/idle.go [hunk 24]
```diff
@@ -266,6 +267,7 @@ func (m *Manager) isClosed() bool {
 	return atomic.LoadInt32(&m.closed) == 1
 }
 
+// Close stops the timer associated with the Manager, if it exists.
```
Comment (arvindbr8) [inline internal/idle/idle.go:270]: Should we say something like?

```suggestion
// Close prevents calls to enter/exit idle immediately.  This call also
// ensures we are not entering/exiting idle mode. 
```

Stoping the timer seems like a side effect?

File: internal/testutils/stats/test_metrics_recorder.go [hunk 25]
```diff
@@ -49,6 +49,7 @@ type TestMetricsRecorder struct {
 	data map[estats.Metric]float64
 }
 
+// NewTestMetricsRecorder returns new TestMetricsRecorder
```
Comment (arvindbr8) [inline internal/testutils/stats/test_metrics_recorder.go:52]:
```suggestion
// NewTestMetricsRecorder returns new TestMetricsRecorder.
```

File: internal/testutils/stats/test_metrics_recorder.go [hunk 26]
```diff
@@ -109,6 +111,9 @@ type MetricsData struct {
 	LabelVals []string
 }
 
+// WaitForInt64Count waits for an int64 count metric to be recorded and
+// verifies that the recorded metrics data matches the expected
+// metricsDataWant.
```
Comment (arvindbr8) [inline internal/testutils/stats/test_metrics_recorder.go:116]:
```suggestion
// WaitForInt64Count waits for an int64 count metric to be recorded and verifies
// that the recorded metrics data matches the expected metricsDataWant.
```

File: internal/testutils/stats/test_metrics_recorder.go [hunk 27]
```diff
@@ -120,6 +125,8 @@ func (r *TestMetricsRecorder) WaitForInt64Count(ctx context.Context, metricsData
 	}
 }
 
+// RecordInt64Count sends the metrics data to the intCountCh channel
+// and updates the internal data map with the recorded value.
```
Comment (arvindbr8) [inline internal/testutils/stats/test_metrics_recorder.go:129]:
```suggestion
// RecordInt64Count sends the metrics data to the intCountCh channel and updates
// the internal data map with the recorded value.
```

File: internal/testutils/stats/test_metrics_recorder.go [hunk 28]
```diff
@@ -133,6 +140,9 @@ func (r *TestMetricsRecorder) RecordInt64Count(handle *estats.Int64CountHandle,
 	r.data[handle.Name] = float64(incr)
 }
 
+// WaitForFloat64Count waits for a float count metric to be recorded and
+// verifies that the recorded metrics data matches the expected
+// metricsDataWant.
```
Comment (arvindbr8) [inline internal/testutils/stats/test_metrics_recorder.go:145]:
```suggestion
// WaitForFloat64Count waits for a float count metric to be recorded and
// verifies that the recorded metrics data matches the expected metricsDataWant.
```

File: internal/testutils/stats/test_metrics_recorder.go [hunk 29]
```diff
@@ -157,6 +169,9 @@ func (r *TestMetricsRecorder) RecordFloat64Count(handle *estats.Float64CountHand
 	r.data[handle.Name] = incr
 }
 
+// WaitForInt64Histo waits for an int histo metric to be recorded and
+// verifies that the recorded metrics data matches the expected
+// metricsDataWant.
```
Comment (arvindbr8) [inline internal/testutils/stats/test_metrics_recorder.go:174]:
```suggestion
// WaitForInt64Histo waits for an int histo metric to be recorded and verifies
// that the recorded metrics data matches the expected metricsDataWant.
```

File: mem/buffers.go [hunk 30]
```diff
@@ -194,19 +197,19 @@ func (b *buffer) read(buf []byte) (int, Buffer) {
 	return n, b
 }
 
-// String returns a string representation of the buffer. May be used for
-// debugging purposes.
 func (b *buffer) String() string {
```
Comment (arvindbr8) [inline mem/buffers.go:200]: mmm, interesting that this one is removed.

File: rpc_util.go [hunk 31]
```diff
@@ -791,9 +791,8 @@ func checkRecvPayload(pf payloadFormat, recvCompress string, haveCompressor bool
 		if !haveCompressor {
 			if isServer {
 				return status.Newf(codes.Unimplemented, "grpc: Decompressor is not installed for grpc-encoding %q", recvCompress)
-			} else {
-				return status.Newf(codes.Internal, "grpc: Decompressor is not installed for grpc-encoding %q", recvCompress)
 			}
+			return status.Newf(codes.Internal, "grpc: Decompressor is not installed for grpc-encoding %q", recvCompress)
```
Comment (arvindbr8) [inline rpc_util.go:795]: This seems unrelated?

File: scripts/revive.toml [hunk 32]
```diff
@@ -25,11 +29,5 @@
...
-[rule.exported]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.redefines-builtin-id]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.package-comments]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.indent-error-flow]
-    Disabled = true # TODO: Enable after existing issues are fixed
+
+
```
Comment (arvindbr8) [inline scripts/revive.toml:33]: nix new line pls

File: scripts/revive.toml [hunk 33]
```diff
@@ -25,11 +29,5 @@
...
 [rule.import-shadowing]
     Disabled = true
-[rule.exported]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.redefines-builtin-id]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.package-comments]
-    Disabled = true # TODO: Enable after existing issues are fixed
-[rule.indent-error-flow]
-    Disabled = true # TODO: Enable after existing issues are fixed
```
Comment (arvindbr8) [inline scripts/revive.toml:35]: Could we keep this PR focused on that change and address the other removals in a separate PR? This will make it easier to track the history of these changes in the future.

**Review by arvindbr8 — APPROVED** [review]
Overall LGTM, modulo a few comments and nits.

File: mem/buffers.go (hunk 30 above)
Comment (purnesh42H) [reply mem/buffers.go:200]: yeah I was removing the docstrings for non exported structs

File: rpc_util.go (hunk 31 above)
Comment (purnesh42H) [reply rpc_util.go:795]: it fixes one lingering `superflous-else`

File: scripts/revive.toml (hunk 33 above)
Comment (purnesh42H) [reply scripts/revive.toml:35]: I think its fine to just enable the remaining ones with this PR since at this time we are sure there are no more existing issues to be fixed

File: scripts/revive.toml (hunk 32 above)
Comment (purnesh42H) [reply scripts/revive.toml:33]: Done

File: internal/channelz/trace.go (hunk 6 above)
Comment (purnesh42H) [reply internal/channelz/trace.go:96]: there is no built-in rule but revive provides framework to write custom rules or we can ask revive contributors for this

File: scripts/revive.toml (hunk 33 above)
Comment (purnesh42H) [reply scripts/revive.toml:35]: Modifying the PR title and merge commit to specify the PR also enable remaining disabled rules
